### DISCUSSION TITLE
Looking for a decent LoRaWan board with full Micropython support

### DISCUSSION THREAD

**Discussion by volkerjaenisch** [micropython/micropython#12169] [Hardware & Peripherals]
I am new to LoRaWan, so please accept patiently my stupid questions.

The goal is to support the small village I live in with low cost soil moisture sensors to get a picture of the current drought level.
And to get this overall picture I need a low voltage long range (5-10km) transmission technology.

This is a non profit project. We all have to adopt to climate change and I like to contribute.

Currently I am working on an [ultra low power ESP32 setup ](https://github.com/orgs/micropython/discussions/12111) for the digital SoMoSe Soil Moisture Sensor. 

With a DFRobot ESP32 Firebeetle the average current is around 0.35mA with one measurement per 5 minutes and WLAN transmission to a MQTT server. For 10 seconds the sensor and the WLAN are drawing 100 mA, but in the following  290 seconds of deep sleep only 0.3 mA is drawn. 

This setup should be easily powered by a 6V 1W solar panel for a long time (buffered by e.g. a 2AH LiPo). Please correct me if I talk nonsense.

Now I like to extend this from WLAN to LoRaWan. The numbers I read on LoRa suggest that it is possible to get much more range on the same power consumption.

I need your advice on two components to play with LoRaWAN:
1) A low cost, low power breakout which can be integrated easily (I2C, UART, SPI, ..) with an arbitrary ESP32 board. 
[Something like that but good ](https://www.amazon.de/Breakout-Antenne-Funkmodul-Arduino-Raspberry/dp/B084NWDFH1)

2) A LoRaWAN Receiver/Gateway to handle the incoming signals and route them to a WLAN.
[LILYGO T-Beam Meshtastic LORA32 868MHz ESP32 Entwicklungsboard WiFi BLE CH9102F Chip TTGO Mainboards Soldered OLED Module ](https://www.amazon.de/LILYGO-Entwicklungsboard-Meshtastic-Mainboards-Soldered/dp/B0B454TB6S)

Please keep in mind that the product you may recommend has to be available in the EU. The EU is quite over regulated. Some parts like the new FeatherS3 (by the way a really masterfully created piece of Art) is not officially available in the EU, due its missing CE certificate.

Any help appreciated

Volker

**Comment by davefes** [comment]
2) Is there a Micropython library that can run on this board as a gateway?

1) I use similar modules using:
https://github.com/Wei1234c/SX127x_driver_for_MicroPython_on_ESP8266/tree/master

He has been quite responsive to issues raised.  I have done 2KM line-of-sight with his default transmit power 2 setting.  For something as infrequent as soil moisture measurement, transmit power 12 shouldn't be too hard on the battery.  Do you need every 5 minutes?

I run several units on 1W panels, but several weeks of cloudy weather can take the battery down to battery protection kick-in point.  

These will probably only be used during the growing season, correct?

**Comment by volkerjaenisch** [comment]
Dear @davefes 


> 2. Is there a Micropython library that can run on this board as a gateway?
> 
>     3. I use similar modules using:
>        https://github.com/Wei1234c/SX127x_driver_for_MicroPython_on_ESP8266/tree/master
> 
> 
> He has been quite responsive to issues raised. I have done 2KM line-of-sight with his default transmit power 2 setting. For something as infrequent as soil moisture measurement, transmit power 12 shouldn't be too hard on the battery. Do you need every 5 minutes?

No. But doing science the shorter the better time resolution.

> 
> I run several units on 1W panels, but several weeks of cloudy weather can take the battery down to battery protection kick-in point.
Good to hear.
> 
> These will probably only be used during the growing season, correct?
Our village does not grow any plants. We have only cattle for milk production and some for meat. We have only 20cm soil and below 1-2 km gravel. So we are really dependent an on surface soil moisture.

Disclosure: 
I am no farmer but a climate scientist, trying help my community. Climate change moves the climate zones at a pace of 18 km/year to the north. In ten years the Saharan climate already devastating Spain, Italiy, Greece, Turkey will be in my village.

My project aims to support my village to get measurements of the current state to make it easier to accept the final truth that cattle need 30 times more water than humans. And in the end one of us has to go - the cattle or us.

Our water resources in the fattest region of Germany are spend. Other regions a facing a even more severe drought in the third year. 

And it will not stop. Even if we stop the CO2 emissions now. It will not stop. CO2 persists 1000 years in the atmosphere. 
So even if we stop this nonsense Coal/Oil-Burning we will have to fight for our existence. But we still have a chance.

But if we do not stop the CO2 emissions there will be nothing left to fight for.

Cheers,
Volker

**Reply by glenn20** [reply]
> > Do you need every 5 minutes?
> 
> No. But doing science the shorter the better time resolution.

One useful approach is to wake up every 5 mins (as the science needs) to take a measurement, but buffer up the measurements and send every hour or so (as the application demands). You can save readings in NVS storage (ESP32) or rtc memory across deepsleep (or flash - at the cost of longevity of your flash storage). This approach can save a lot of power whatever transmission technology you use.

**Reply by volkerjaenisch** [reply]
@glenn20 
Thank you for this really good idea. This can help. 
But the sensor current is IMHO the limiting factor.

Cheers,
Volker

**Comment by davefes** [comment]
I run 8 head of cattle here and during the winter moisture is not a problem ... we have had 300-400mm of the last 3-4 weeks.  We need the sun!

Are they for beef or milking?

Several years ago I was "following" a French guy, doing some work in Africa on complete LoRaWAN systems for agriculture.  Because I haven't done a specific LoRaWAN gateway I think you need to work out the higher system level details first.  Maybe, the 2nd link to Arduino code on those units might be easier.   Search Github uPyLoRaWAN-master, however in my brief look I don't see anything gateway-related. 

https://www.instructables.com/Arduino-LoRa-ESP32-LoRa-Gateway-Obtain-Remote-Ardu/
has a link to:
https://github.com/Makerfabs/Makepython-Lora-Gateway

Good luck!

**Reply by davefes** [reply]
That last link I recognise a lot of Wei's code base in there.  Something that is not clear to me is how you run several networks in the same area.  The 433MHz modules I am using appear to be only set on 433 or 434MHz.  How many channels can you use in the EU at 800-900MHz?  I thought you could change sync_word to many different values but it appears that 0x12 and 0x34 are the common ones that work.

Again, you need a clear understanding of the higher-level stuff, ie how to cope with other systems in your area.  Have you watched the Spiess Youtube videos?

**Reply by davefes** [reply]
In anticipation of a thread on powering your sensors I had a look for some low power PFETs.  There is  a wide-range of devices from Vishay, ie the Si2347DS looks promising.

I would switch the PFET with something like a BC548.  If you can find guaranteed logic threshold levels for the ESP32 GPIO you might even be able to switch the Si2347DS directly from GPIO.

**Reply by volkerjaenisch** [reply]
Mh. Looks like I was a bit naive.
I thought LoRaWan is a well established technology. But @all your feedback suggests that this is all quite experimental. 
I will do some reading, watching and learning first.
Cheers,
Volker

**Reply by davefes** [reply]
"Experimental" to me.  Read through this last night:
https://os.mbed.com/docs/mbed-os/v6.16/apis/LoRa-tutorial.html
I have two LoRa links for water tank monitoring, one for PV panel control and another for microhydro control.  All 4 local units are really only looking at one remote, however they also connect to the internet.

Dealing with multiple sensors and using a concentrator would be new territory for me.

**Comment by volkerjaenisch** [comment]
> I run 8 head of cattle here and during the winter moisture is not a problem ... we have had 300-400mm of the last 3-4 weeks. We need the sun!

Ground water accumulates usually during the winter. The snows slowly melting does the trick. But we have no snow any more. The glaciers in the mountains are nearly gone.

Rainfall is either sparse or it comes as a flood. Goes straight into the river. A guy from the lots of power stations we have along our river told me: They got 500 m³/sec at one of the last rainfalls. This was twice the capacity the dams are made for. 

It will be a hard ride. And we see just the beginning.

Cheers,
Volker

**Reply by billbreit** [reply]
Hi Volker,

I am working on a 'smart' control unit for solar-powered water pumps in remote areas, primarily to serve wildlife. 

The control unit would try to maximize the liters pumped while minimizing the number of on-off stresses to the pump on cloudy days, perhaps with a RP Zero 'mainframe' for data analysis and learning.  It would run during sunny periods and be turned off most of the time ( by a continuously running RP Pico ).  The installation would have moisture sensors.

I've concluded that logging moisture data is better than broadcasting it via radio, LoRa being the leading candidate.  Sampling a couple of times a day and logging the data is probably sufficient.  Every few weeks or even months, someone could trek out to the various sites and collect the data via WiFi.  The data is not immediately actionable - it's for monitoring.

Maybe rethink you requirements.  Perhaps do the LoRa as phase two of the project.  

Brother In Arms

**Comment by robert-hh** [comment]
> I thought LoRaWan is a well established technology. But ALL your feedback suggests that this is all quite experimental.

There is/was a company called Pycom, which had a set of ESP32 boards with Sequans LoRa hardware and MicroPython as software base. These work quite well with simple LoRa and LoRa WAN. Unfortunately, Pycom went bankrupt due to some unfortunate business decisions, and the company that purchased the assets does not continue the ESP32 series of products. The ESP32 boards were quite generic, and I ported the Pycom firmware to 2 different Heltec boards, the Wireless Stick and the Wireless Stick Lite. So it should be possible to transfer their firmware to a ESP32 WROOM or WROVER type board with SX1276 LoRa hardware. Their MicroPython is similar to the Genuine Micropython, more than CircuitPython. It was created by Daniel Campora, who made the CC3200 port.

LoRa is established, and if you go e.g. to the TTN web site, you will find links to MANY products, both nodes and gateways. For your application, using LoRa WAN and a gateway with a good antenna located in the middle of the to-be-covered area seems to be a good approach. You can then use the TTN network to collect and transfer the data.

**Comment by robert-hh** [comment]
About distances. With a ESP32 based Node I drove around the city and checked, where I can get messages passed both to a gateway at home and other gateways in the area. To the one at home I get contact up to a distance of ~2300m.  It depends a lot on the area and building. But in most cases the signal of the node were picked up by gateways 7 or 8 km apart. 
Being curious I drove to these places, and one had a high gain antennas placed on the roof of a house in a height of about 20 m, and the other was on the top of a hill with a good antenna with a clear line of sight to most places where I tested. Not surprising: physics always rules.
